leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java

```diff
@@ -0,0 +1,232 @@
+        byte[] data = serialize(client);
+        try (Jedis j = pool.getResource()) {
+            byte[] k = (EP_CLIENT + client.getEndpoint()).getBytes();
+            byte[] old = j.getSet(k, data);
+
+            j.expire(k, client.getLifeTimeInSec().intValue());
+
+            // secondary index
+            byte[] idx = (REG_EP + client.getRegistrationId()).getBytes();
+            j.set(idx, client.getEndpoint().getBytes());
+
+            j.expire(idx, client.getLifeTimeInSec().intValue());
+
+            if (old != null) {
+                Client oldClient = deserialize(old);
+                for (ClientRegistryListener l : listeners) {
+                    l.unregistered(oldClient);
+                }
+            }
+            for (ClientRegistryListener l : listeners) {
+                l.registered(client);
+            }
+            return true;
+        }
+    }
+
+    @Override
+    public Client updateClient(ClientUpdate update) {
+        try (Jedis j = pool.getResource()) {
+            // fetch the client ep by registration ID index
+            byte[] key = j.get((REG_EP + update.getRegistrationId()).getBytes());
+            if (key == null) {
+                return null;
+            }
+
+            // fetch the client
+            byte[] data = j.get(ByteArrayUtils.concatenate(EP_CLIENT.getBytes(), key));
+            if (data == null) {
+                return null;
+            }
+
+            Client c = deserialize(data);
+            Client clientUpdated = update.updateClient(c);
+            // store the new client
+            byte[] k = (EP_CLIENT + clientUpdated.getEndpoint()).getBytes();
+
+            j.setex(k, clientUpdated.getLifeTimeInSec().intValue(), serialize(clientUpdated));
+
+            byte[] idx = (REG_EP + clientUpdated.getRegistrationId()).getBytes();
+
+            j.expire(idx, clientUpdated.getLifeTimeInSec().intValue());
+
+            // notify listener
+            for (ClientRegistryListener l : listeners) {
+                l.updated(clientUpdated);
+            }
+            return clientUpdated;
+        }
+    }
+
+    @Override
+    public Client deregisterClient(String registrationId) {
+        try (Jedis j = pool.getResource()) {
+            // first rename for atomicity
+            if (!"OK".equals(j.rename((REG_EP + registrationId).getBytes(),
+                    ("TODELETE#" + REG_EP + registrationId).getBytes()))) {
+                System.err.println("del meh");
+                return null;
+            }
+
+            // fetch the client ep by registration ID index
+            byte[] key = j.get(("TODELETE#" + REG_EP + registrationId).getBytes());
+            if (key == null) {
+                return null;
+            }
+            byte[] data = j.get(ByteArrayUtils.concatenate(EP_CLIENT.getBytes(), key));
+
+            Client c = deserialize(data);
+
+            // delete everything
+            j.del(("TODELETE#" + REG_EP + registrationId).getBytes());
```

Comment:
`("TODELETE#" + REG_EP + registrationId).getBytes() ` -> This key is built 3 times in this method.